src/DurableTask.Core/Tracing/TraceHelper.cs

```diff
@@ -249,18 +266,99 @@ internal static void EmitActivityforTaskFinished(OrchestrationInstance? instance
             {
-                newActivity.SetTag("otel.status_code", "ERROR");
+                string statusDescription = "";
 
-                FailureDetails? failureDetails = taskFailedEvent.FailureDetails;
-                if (failureDetails != null)
+                if (errorPropagationMode == ErrorPropagationMode.SerializeExceptions)
                 {
-                    newActivity.SetTag("otel.status_description", failureDetails.ErrorMessage);
+                    statusDescription = JsonDataConverter.Default.Deserialize<Exception>(taskFailedEvent.Details).Message;
                 }
+                else if (errorPropagationMode == ErrorPropagationMode.UseFailureDetails)
+                {
+                    FailureDetails? failureDetails = taskFailedEvent.FailureDetails;
+                    if (failureDetails != null)
+                    {
+                        statusDescription = failureDetails.ErrorMessage;
+                    }
+                }
+
+                newActivity.SetStatus(ActivityStatusCode.Error, statusDescription);
             }
 
             newActivity.Dispose();
         }
 
+        internal static Activity? CreateTraceActivityForTaskFinished(
+            OrchestrationInstance? instance,
+            TaskScheduledEvent taskScheduledEvent)
+        {
+            if (taskScheduledEvent == null)
+            {
+                return null;
+            }
+
+            if (!taskScheduledEvent.TryGetParentTraceContext(out ActivityContext activityContext))
+            {
+                return null;
+            }
+
+            Activity? newActivity = ActivityTraceSource.StartActivity(
+                name: taskScheduledEvent.Name ?? "",
+                kind: ActivityKind.Client,
+                startTime: taskScheduledEvent.Timestamp,
+                parentContext: activityContext);
+
+            if (newActivity == null)
+            {
+                return null;
+            }
+
+            newActivity.AddTag("dtfx.type", "activity");
+            newActivity.AddTag("dtfx.instance_id", instance?.InstanceId);
+            newActivity.AddTag("dtfx.execution_id", instance?.ExecutionId);
+
+            return newActivity;
+        }
+
+        internal static void EmitTraceActivityForTaskCompleted(
+            OrchestrationInstance? orchestrationInstance,
+            TaskScheduledEvent taskScheduledEvent)
+        {
+            Activity? activity = CreateTraceActivityForTaskFinished(orchestrationInstance, taskScheduledEvent);
+
+            activity?.Dispose();
+        }
+
+        internal static void EmitTraceActivityForTaskFailed(
+            OrchestrationInstance? orchestrationInstance,
+            TaskScheduledEvent taskScheduledEvent,
+            TaskFailedEvent? taskFailedEvent,
+            ErrorPropagationMode errorPropagationMode)
+        {
+            Activity? activity = CreateTraceActivityForTaskFinished(orchestrationInstance, taskScheduledEvent);
```

Comment:
nit: recommend checking for null and returning right away:

```suggestion
            Activity? activity = CreateTraceActivityForTaskFinished(orchestrationInstance, taskScheduledEvent);
            
            if (activity is null)
            {
                return;
            }
```